ocaml/xapi/xapi_vdi_helpers.ml

```
@@ -23,6 +23,87 @@ open Threadext
+		Eq (Field "VDI", Literal (Ref.string_of vdi)),
+		Eq (Field "currently_attached", Literal "true"))
+	) in
+	let (vbd, device) =
+		match vbds with
+		| vbd :: _ -> (vbd, Db.VBD.get_device ~__context ~self:vbd)
+		| [] -> begin
+			let dom0 = Helpers.get_domain_zero ~__context in
+			let vbd = Helpers.call_api_functions ~__context (fun rpc session_id ->
+				Client.VBD.create ~rpc ~session_id ~vM:dom0 ~vDI:vdi ~mode:`RW ~_type:`Disk
+				~empty:false ~userdevice:"autodetect" ~bootable:false ~unpluggable:true
+				~qos_algorithm_type:"" ~qos_algorithm_params:[] ~other_config:[] )
+			in
+			Helpers.call_api_functions ~__context (fun rpc session_id ->
+				Client.VBD.plug ~rpc ~session_id ~self:vbd);
+			(vbd, Db.VBD.get_device ~__context ~self:vbd)
+		end
+	in
+	(vbd, device)
+
+(* Detach the rrd-stats vdi *)
+let detach_rrd_vdi ~__context ~vbd =
+	Helpers.call_api_functions ~__context (fun rpc session_id ->
+		Client.VBD.unplug ~rpc ~session_id ~self:vbd);
+	Helpers.call_api_functions ~__context (fun rpc session_id ->
+		Client.VBD.destroy ~rpc ~session_id ~self:vbd)
+
+(* Construct the CStruct for the rrd-stats vdi *)
+let open_rrd_file ~device ~mode =
+	let path = "/dev/" ^ device in
+	let fd = Unix.openfile path mode 0 in
+	let size = 4194304 in
+	let mapping = Bigarray.(Array1.map_file fd char c_layout true size) in
+	Unix.close fd;
+	let cstruct = Cstruct.of_bigarray mapping in
+	cstruct
+
+let format_rrd_vdi ~cstruct ~magic_number =
+	let version = 1 and length = 0 in
+	Cstruct.BE.set_uint32 cstruct 0 magic_number;
+	Cstruct.BE.set_uint32 cstruct 4 (Int32.of_int version);
+	Cstruct.BE.set_uint32 cstruct 8 (Int32.of_int length)
+
+(* Write the rrd stats to the rrd-stats vdi *)
+let write_rrd ~__context ~sr ~vdi ~text =
+	let magic_number = (Int32.of_int 0x7ada7ada) in
+	let (vbd, device) = get_rrd_vdi_device ~__context ~vdi in
+	let cstruct = open_rrd_file ~device ~mode:[Unix.O_RDWR] in
+	if (Cstruct.BE.get_uint32 cstruct 0) <>  magic_number then
```

I don't think the accessor function should format the RRD if the magic number is not right.   You should format the RRD when you create it, and if the magic number is not right when you try to read it you should raise an error to be handled by the caller.